flask/app/gene_viewer.py

```diff
@@ -0,0 +1,250 @@
+                == models.users_groups_table.columns.group_id,
+            )
+            .options(
+                contains_eager(models.Gene.variants)
+                .contains_eager(models.Variant.genotype)
+                .contains_eager(models.Genotype.analysis)
+                .contains_eager(models.Analysis.datasets)
+                .contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters), models.users_groups_table.columns.user_id == user_id)
+        )
+
+    else:
+        app.logger.debug("Processing query - unrestricted based on user id.")
+        query = (
+            models.Gene.query.join(models.Gene.variants)
+            .join(models.Variant.genotype)
+            .join(models.Genotype.analysis, models.Genotype.dataset)
+            .join(models.Dataset.tissue_sample)
+            .join(models.TissueSample.participant)
+            .join(models.Participant.family)
+            .options(
+                contains_eager(models.Gene.variants)
+                .contains_eager(models.Variant.genotype)
+                .contains_eager(models.Genotype.analysis)
+                .contains_eager(models.Analysis.datasets)
+                .contains_eager(models.Dataset.tissue_sample)
+                .contains_eager(models.TissueSample.participant)
+                .contains_eager(models.Participant.family),
+            )
+            .filter(or_(*filters))
+        )
+
+    # defaults to json unless otherwise specified
+    app.logger.info(request.accept_mimetypes)
+
+    if request.accept_mimetypes["application/json"]:
+
+        app.logger.info("Defaulting to json response")
+
+        q = query.all()
+
+        if len(q) == 0:
+            app.logger.error("No requested genes were found.")
+            abort(400, description="No requested genes were found.")
+        elif len(q) < len(genes):
+            app.logger.error("Not all requested genes were found.")
+            abort(400, description="Not all requested genes were found.")
+
+        return jsonify(
+            [
+                {
+                    **asdict(gene),
+                    "variants": [
+                        {
+                            **asdict(variant),
+                            "genotype": [
+                                {
+                                    **asdict(genotype),
+                                    "participant_codename": genotype.dataset.tissue_sample.participant.participant_codename,
+                                }
+                                for genotype in variant.genotype
+                            ],
+                        }
+                        for variant in gene.variants
+                    ],
+                }
+                for gene in q
+            ],
+        )
+    elif request.accept_mimetypes["text/csv"]:
```

Comment:
Since we're using this approach, we need a final else clause to return 406, or there's a branch where this function returns None implicitly